Random unrecoverable "bucketMetadata" is null error

Using the docker image, sometimes we're seeing this error randomly and the container will die and  fail to restart:

```
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'bucketStore' defined in class path resource [com/adobe/testing/s3mock/store/StoreConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.adobe.testing.s3mock.store.BucketStore]: Factory method 'bucketStore' threw exception; nested exception is java.lang.NullPointerException: Cannot invoke "com.adobe.testing.s3mock.store.BucketMetadata.getObjects()" because "bucketMetadata" is null
```

The error is unrecoverable and we have to start over with the container (deleting all files), which is obviously a laborious process. It's happened a few times now. 

Here's our `docker-compose.yml`:

```yaml
version: '3.8'
services:
  s3-local:
    image: adobe/s3mock:latest
    environment:
      - initialBuckets=bucket-1,bucket-2
      - retainFilesOnExit=true
      - root=/var/lib/s3mock/data
    ports:
      - '9090:9090' #http
      - '9191:9191' #https
    volumes:
      - './docker/s3mock:/var/lib/s3mock/data'
```

Each of the buckets has a `bucketMetadata.json` present, which looks to be valid (no signs of corruption or similar).

Got the same error (with `retainFilesOnExit=true`) every time the s3mock container is stopped, removed and started again.

docker-compose.yml:

```yml
services:
  s3mock:
    image: adobe/s3mock:3.0.0
    environment:
      - initialBuckets=bucket1
      - root=/data
    ports:
      - 9090:9090
    volumes:
      - ./data:/data
```

Steps to produce error:

```bash
docker compose up -d

mkdir ~/.aws
cat <<EOF > ~/.aws/credentials
[default]
aws_access_key_id = dummy
aws_secret_access_key = dummy
EOF

mkdir /tmp/test
truncate -s 1M /tmp/test/file.dat

aws --endpoint-url http://127.0.0.1:9090 s3 sync /tmp/test s3://bucket1/
aws --endpoint-url http://127.0.0.1:9090 s3 ls s3://bucket1/

docker compose logs
# start up success, capture logs

docker compose down
docker compose up -d

docker compose logs
# start up failed, capture logs again

sudo rm -r data/bucket1
docker compose up -d
# start up success again
```

<details><summary>1st container log:</summary>

```
s3mock-s3mock-1  | OpenJDK 64-Bit Server VM warning: Ignoring option --illegal-access=warn; support was removed in 17.0
s3mock-s3mock-1  |
s3mock-s3mock-1  |
s3mock-s3mock-1  |
s3mock-s3mock-1  |
s3mock-s3mock-1  |   .-')              _   .-')                           .-. .-')
s3mock-s3mock-1  |  ( OO ).           ( '.( OO )_                         \  ( OO )
s3mock-s3mock-1  | (_)---\_) .-----.   ,--.   ,--.).-'),-----.    .-----. ,--. ,--.
s3mock-s3mock-1  | /    _ | /  -.   \  |   `.'   |( OO'  .-.  '  '  .--./ |  .'   /
s3mock-s3mock-1  | \  :` `. '-' _'  |  |         |/   |  | |  |  |  |('-. |      /,
s3mock-s3mock-1  |  '..`''.)   |_  <   |  |'.'|  |\_) |  |\|  | /_) |OO  )|     ' _)
s3mock-s3mock-1  | .-._)   \.-.  |  |  |  |   |  |  \ |  | |  | ||  |`-'| |  .   \
s3mock-s3mock-1  | \       /\ `-'   /  |  |   |  |   `'  '-'  '(_'  '--'\ |  |\   \
s3mock-s3mock-1  |  `-----'  `----''   `--'   `--'     `-----'    `-----' `--' '--'
s3mock-s3mock-1  |
s3mock-s3mock-1  |
s3mock-s3mock-1  |
s3mock-s3mock-1  |
s3mock-s3mock-1  |
s3mock-s3mock-1  |
s3mock-s3mock-1  | 2023-07-25T06:56:49.258Z  INFO 1 --- [           main] c.a.testing.s3mock.S3MockApplication     : Starting S3MockApplication using Java 17.0.7 with PID 1 (/s3moc
k.jar started by root in /)
s3mock-s3mock-1  | 2023-07-25T06:56:49.262Z  INFO 1 --- [           main] c.a.testing.s3mock.S3MockApplication     : No active profile set, falling back to 1 default profile: "defa
ult"
s3mock-s3mock-1  | 2023-07-25T06:56:51.046Z  INFO 1 --- [           main] o.s.b.w.e.j.JettyServletWebServerFactory : Server initialized with port: 9191
s3mock-s3mock-1  | 2023-07-25T06:56:51.212Z  INFO 1 --- [           main] org.eclipse.jetty.server.Server          : jetty-11.0.15; built: 2023-04-11T18:37:53.775Z; git: 5bc5e562c8
d05c5862505aebe5cf83a61bdbcb96; jvm 17.0.7+7-alpine-r1
s3mock-s3mock-1  | 2023-07-25T06:56:51.276Z  INFO 1 --- [           main] o.e.j.s.h.ContextHandler.application     : Initializing Spring embedded WebApplicationContext
s3mock-s3mock-1  | 2023-07-25T06:56:51.277Z  INFO 1 --- [           main] w.s.c.ServletWebServerApplicationContext : Root WebApplicationContext: initialization completed in 1903 ms
s3mock-s3mock-1  | 2023-07-25T06:56:51.544Z  INFO 1 --- [           main] o.e.j.s.session.DefaultSessionIdManager  : Session workerName=node0
s3mock-s3mock-1  | 2023-07-25T06:56:51.553Z  INFO 1 --- [           main] o.e.jetty.server.handler.ContextHandler  : Started o.s.b.w.e.j.JettyEmbeddedWebAppContext@3dddbe65{applica
tion,/,[file:///tmp/jetty-docbase.9191.8440047267480645554/],AVAILABLE}
s3mock-s3mock-1  | 2023-07-25T06:56:51.561Z  INFO 1 --- [           main] org.eclipse.jetty.server.Server          : Started Server@6c67e137{STARTING}[11.0.15,sto=0] @3569ms
s3mock-s3mock-1  | 2023-07-25T06:56:51.580Z  INFO 1 --- [           main] c.a.t.s3mock.store.StoreConfiguration    : Using existing folder "/data" as root folder. Will retain files
 on exit: false
s3mock-s3mock-1  | 2023-07-25T06:56:51.687Z  INFO 1 --- [           main] c.a.t.s3mock.store.StoreConfiguration    : Creating initial bucket bucket1.
s3mock-s3mock-1  | 2023-07-25T06:56:52.089Z  WARN 1 --- [           main] i.m.c.i.binder.jvm.JvmGcMetrics          : GC notifications will not be available because com.sun.manageme
nt.GarbageCollectionNotificationInfo is not present
s3mock-s3mock-1  | 2023-07-25T06:56:52.160Z  INFO 1 --- [           main] o.s.b.a.e.web.EndpointLinksResolver      : Exposing 1 endpoint(s) beneath base path '/actuator'
s3mock-s3mock-1  | 2023-07-25T06:56:52.229Z  INFO 1 --- [           main] o.e.j.s.h.ContextHandler.application     : Initializing Spring DispatcherServlet 'dispatcherServlet'
s3mock-s3mock-1  | 2023-07-25T06:56:52.229Z  INFO 1 --- [           main] o.s.web.servlet.DispatcherServlet        : Initializing Servlet 'dispatcherServlet'
s3mock-s3mock-1  | 2023-07-25T06:56:52.230Z  INFO 1 --- [           main] o.s.web.servlet.DispatcherServlet        : Completed initialization in 1 ms
s3mock-s3mock-1  | 2023-07-25T06:56:52.241Z  INFO 1 --- [           main] o.e.jetty.util.ssl.SslContextFactory     : x509=X509@2b0b4d53(selfsigned,h=[localhost, adobe s3mock],a=[],
w=[]) for Server@7068f7ca[provider=null,keyStore=null,trustStore=null]
s3mock-s3mock-1  | 2023-07-25T06:56:52.375Z  INFO 1 --- [           main] o.e.jetty.server.AbstractConnector       : Started SslValidatingServerConnector@117e0fe5{SSL, (ssl, http/1
.1)}{0.0.0.0:9191}
s3mock-s3mock-1  | 2023-07-25T06:56:52.397Z  INFO 1 --- [           main] o.e.jetty.server.AbstractConnector       : Started ServerConnector@2a65bb85{HTTP/1.1, (http/1.1)}{0.0.0.0:
9090}
s3mock-s3mock-1  | 2023-07-25T06:56:52.400Z  INFO 1 --- [           main] o.s.b.web.embedded.jetty.JettyWebServer  : Jetty started on port(s) 9191 (ssl, http/1.1), 9090 (http/1.1)
with context path '/'
s3mock-s3mock-1  | 2023-07-25T06:56:52.421Z  INFO 1 --- [           main] c.a.testing.s3mock.S3MockApplication     : Started S3MockApplication in 3.796 seconds (process running for
 4.429)
s3mock-s3mock-1  | 2023-07-25T07:03:42.414Z  WARN 1 --- [qtp962287291-48] ocalVariableTableParameterNameDiscoverer : Using deprecated '-debug' fallback for parameter name resolutio
n. Compile the affected code with '-parameters' instead or avoid its introspection: com.adobe.testing.s3mock.BucketController
s3mock-s3mock-1  | 2023-07-25T07:03:42.478Z  WARN 1 --- [qtp962287291-49] ocalVariableTableParameterNameDiscoverer : Using deprecated '-debug' fallback for parameter name resolutio
n. Compile the affected code with '-parameters' instead or avoid its introspection: com.adobe.testing.s3mock.ObjectController
```
</details>

<details><summary>2nd container log:</summary>

```
s3mock-s3mock-1  | OpenJDK 64-Bit Server VM warning: Ignoring option --illegal-access=warn; support was removed in 17.0
s3mock-s3mock-1  |
s3mock-s3mock-1  |
s3mock-s3mock-1  |
s3mock-s3mock-1  |
s3mock-s3mock-1  |   .-')              _   .-')                           .-. .-')
s3mock-s3mock-1  |  ( OO ).           ( '.( OO )_                         \  ( OO )
s3mock-s3mock-1  | (_)---\_) .-----.   ,--.   ,--.).-'),-----.    .-----. ,--. ,--.
s3mock-s3mock-1  | /    _ | /  -.   \  |   `.'   |( OO'  .-.  '  '  .--./ |  .'   /
s3mock-s3mock-1  | \  :` `. '-' _'  |  |         |/   |  | |  |  |  |('-. |      /,
s3mock-s3mock-1  |  '..`''.)   |_  <   |  |'.'|  |\_) |  |\|  | /_) |OO  )|     ' _)
s3mock-s3mock-1  | .-._)   \.-.  |  |  |  |   |  |  \ |  | |  | ||  |`-'| |  .   \
s3mock-s3mock-1  | \       /\ `-'   /  |  |   |  |   `'  '-'  '(_'  '--'\ |  |\   \
s3mock-s3mock-1  |  `-----'  `----''   `--'   `--'     `-----'    `-----' `--' '--'
s3mock-s3mock-1  |
s3mock-s3mock-1  |
s3mock-s3mock-1  |
s3mock-s3mock-1  |
s3mock-s3mock-1  |
s3mock-s3mock-1  |
s3mock-s3mock-1  | 2023-07-25T07:05:08.299Z  INFO 1 --- [           main] c.a.testing.s3mock.S3MockApplication     : Starting S3MockApplication using Java 17.0.7 with PID 1 (/s3moc
k.jar started by root in /)
s3mock-s3mock-1  | 2023-07-25T07:05:08.303Z  INFO 1 --- [           main] c.a.testing.s3mock.S3MockApplication     : No active profile set, falling back to 1 default profile: "defa
ult"
s3mock-s3mock-1  | 2023-07-25T07:05:09.993Z  INFO 1 --- [           main] o.s.b.w.e.j.JettyServletWebServerFactory : Server initialized with port: 9191
s3mock-s3mock-1  | 2023-07-25T07:05:10.229Z  INFO 1 --- [           main] org.eclipse.jetty.server.Server          : jetty-11.0.15; built: 2023-04-11T18:37:53.775Z; git: 5bc5e562c8
d05c5862505aebe5cf83a61bdbcb96; jvm 17.0.7+7-alpine-r1
s3mock-s3mock-1  | 2023-07-25T07:05:10.298Z  INFO 1 --- [           main] o.e.j.s.h.ContextHandler.application     : Initializing Spring embedded WebApplicationContext
s3mock-s3mock-1  | 2023-07-25T07:05:10.300Z  INFO 1 --- [           main] w.s.c.ServletWebServerApplicationContext : Root WebApplicationContext: initialization completed in 1886 ms
s3mock-s3mock-1  | 2023-07-25T07:05:10.567Z  INFO 1 --- [           main] o.e.j.s.session.DefaultSessionIdManager  : Session workerName=node0
s3mock-s3mock-1  | 2023-07-25T07:05:10.576Z  INFO 1 --- [           main] o.e.jetty.server.handler.ContextHandler  : Started o.s.b.w.e.j.JettyEmbeddedWebAppContext@36ac8a63{applica
tion,/,[file:///tmp/jetty-docbase.9191.11345237632191431542/],AVAILABLE}
s3mock-s3mock-1  | 2023-07-25T07:05:10.592Z  INFO 1 --- [           main] org.eclipse.jetty.server.Server          : Started Server@53ab0286{STARTING}[11.0.15,sto=0] @3599ms
s3mock-s3mock-1  | 2023-07-25T07:05:10.611Z  INFO 1 --- [           main] c.a.t.s3mock.store.StoreConfiguration    : Using existing folder "/data" as root folder. Will retain files
 on exit: false
s3mock-s3mock-1  | 2023-07-25T07:05:10.650Z  INFO 1 --- [           main] c.a.testing.s3mock.store.BucketStore     : Loading existing bucket bucket1.
s3mock-s3mock-1  | 2023-07-25T07:05:10.653Z  WARN 1 --- [           main] ConfigServletWebServerApplicationContext : Exception encountered during context initialization - cancellin
g refresh attempt: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'fileStoreController' defined in class path resource [com/adobe/t
esting/s3mock/S3MockConfiguration.class]: Unsatisfied dependency expressed through method 'fileStoreController' parameter 0: Error creating bean with name 'objectService' defined i
n class path resource [com/adobe/testing/s3mock/service/ServiceConfiguration.class]: Unsatisfied dependency expressed through method 'objectService' parameter 0: Error creating bea
n with name 'bucketStore' defined in class path resource [com/adobe/testing/s3mock/store/StoreConfiguration.class]: Failed to instantiate [com.adobe.testing.s3mock.store.BucketStor
e]: Factory method 'bucketStore' threw exception with message: Cannot invoke "com.adobe.testing.s3mock.store.BucketMetadata.objects()" because "bucketMetadata" is null
s3mock-s3mock-1  | 2023-07-25T07:05:10.656Z  INFO 1 --- [           main] org.eclipse.jetty.server.Server          : Stopped Server@53ab0286{STOPPING}[11.0.15,sto=0]
s3mock-s3mock-1  | 2023-07-25T07:05:10.658Z  INFO 1 --- [           main] o.e.jetty.server.handler.ContextHandler  : Stopped o.s.b.w.e.j.JettyEmbeddedWebAppContext@36ac8a63{applica
tion,/,[file:///tmp/jetty-docbase.9191.11345237632191431542/],STOPPED}
s3mock-s3mock-1  | 2023-07-25T07:05:10.678Z  INFO 1 --- [           main] .s.b.a.l.ConditionEvaluationReportLogger :
s3mock-s3mock-1  |
s3mock-s3mock-1  |
s3mock-s3mock-1  | Error starting ApplicationContext. To display the condition evaluation report re-run your application with 'debug' enabled.
s3mock-s3mock-1  | 2023-07-25T07:05:10.704Z ERROR 1 --- [           main] o.s.boot.SpringApplication               : Application run failed
s3mock-s3mock-1  |
s3mock-s3mock-1  |
s3mock-s3mock-1  | org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'fileStoreController' defined in class path resource [com/adobe/t
esting/s3mock/S3MockConfiguration.class]: Unsatisfied dependency expressed through method 'fileStoreController' parameter 0: Error creating bean with name 'objectService' defined i
n class path resource [com/adobe/testing/s3mock/service/ServiceConfiguration.class]: Unsatisfied dependency expressed through method 'objectService' parameter 0: Error creating bea
n with name 'bucketStore' defined in class path resource [com/adobe/testing/s3mock/store/StoreConfiguration.class]: Failed to instantiate [com.adobe.testing.s3mock.store.BucketStor
e]: Factory method 'bucketStore' threw exception with message: Cannot invoke "com.adobe.testing.s3mock.store.BucketMetadata.objects()" because "bucketMetadata" is null
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:800) ~[spring-beans-6.0.9.jar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:550) ~[spring-beans-6.0.9.jar!/:6.0.
9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1332)
~[spring-beans-6.0.9.jar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1162) ~[spring-be
ans-6.0.9.jar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:560) ~[spring-beans-6.0
.9.jar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:520) ~[spring-beans-6.0.9
.jar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:326) ~[spring-beans-6.0.9.jar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234) ~[spring-beans-6.0.9.jar!/:6.0
.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:324) ~[spring-beans-6.0.9.jar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:200) ~[spring-beans-6.0.9.jar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:973) ~[spring-beans-6.0.9.j
ar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:941) ~[spring-context-6.0.
9.jar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:608) ~[spring-context-6.0.9.jar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:146) ~[spring-boot-3.1.0.
jar!/:3.1.0]
s3mock-s3mock-1  |      at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:733) ~[spring-boot-3.1.0.jar!/:3.1.0]
s3mock-s3mock-1  |      at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:435) ~[spring-boot-3.1.0.jar!/:3.1.0]
s3mock-s3mock-1  |      at org.springframework.boot.SpringApplication.run(SpringApplication.java:311) ~[spring-boot-3.1.0.jar!/:3.1.0]
s3mock-s3mock-1  |      at org.springframework.boot.builder.SpringApplicationBuilder.run(SpringApplicationBuilder.java:150) ~[spring-boot-3.1.0.jar!/:3.1.0]
s3mock-s3mock-1  |      at com.adobe.testing.s3mock.S3MockApplication.start(S3MockApplication.java:199) ~[classes!/:na]
s3mock-s3mock-1  |      at com.adobe.testing.s3mock.S3MockApplication.start(S3MockApplication.java:163) ~[classes!/:na]
s3mock-s3mock-1  |      at com.adobe.testing.s3mock.S3MockApplication.main(S3MockApplication.java:152) ~[classes!/:na]
s3mock-s3mock-1  |      at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
s3mock-s3mock-1  |      at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source) ~[na:na]
s3mock-s3mock-1  |      at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[na:na]
s3mock-s3mock-1  |      at java.base/java.lang.reflect.Method.invoke(Unknown Source) ~[na:na]
s3mock-s3mock-1  |      at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:49) ~[s3mock.jar:na]
s3mock-s3mock-1  |      at org.springframework.boot.loader.Launcher.launch(Launcher.java:95) ~[s3mock.jar:na]
s3mock-s3mock-1  |      at org.springframework.boot.loader.Launcher.launch(Launcher.java:58) ~[s3mock.jar:na]
s3mock-s3mock-1  |      at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:65) ~[s3mock.jar:na]
s3mock-s3mock-1  | Caused by: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'objectService' defined in class path resource [com/ad
obe/testing/s3mock/service/ServiceConfiguration.class]: Unsatisfied dependency expressed through method 'objectService' parameter 0: Error creating bean with name 'bucketStore' def
ined in class path resource [com/adobe/testing/s3mock/store/StoreConfiguration.class]: Failed to instantiate [com.adobe.testing.s3mock.store.BucketStore]: Factory method 'bucketSto
re' threw exception with message: Cannot invoke "com.adobe.testing.s3mock.store.BucketMetadata.objects()" because "bucketMetadata" is null
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:800) ~[spring-beans-6.0.9.jar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:550) ~[spring-beans-6.0.9.jar!/:6.0.
9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1332)
~[spring-beans-6.0.9.jar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1162) ~[spring-be
ans-6.0.9.jar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:560) ~[spring-beans-6.0
.9.jar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:520) ~[spring-beans-6.0.9
.jar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:326) ~[spring-beans-6.0.9.jar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234) ~[spring-beans-6.0.9.jar!/:6.0
.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:324) ~[spring-beans-6.0.9.jar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:200) ~[spring-beans-6.0.9.jar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:254) ~[spring-beans-6.0.9.jar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1417) ~[spring-beans-6.0.9.jar!/
:6.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1337) ~[spring-beans-6.0.9.jar!/:6
.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:888) ~[spring-beans-6.0.9.jar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:791) ~[spring-beans-6.0.9.jar!/:6.0.9]
s3mock-s3mock-1  |      ... 28 common frames omitted
s3mock-s3mock-1  | Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'bucketStore' defined in class path resource [com/adobe/testing
/s3mock/store/StoreConfiguration.class]: Failed to instantiate [com.adobe.testing.s3mock.store.BucketStore]: Factory method 'bucketStore' threw exception with message: Cannot invok
e "com.adobe.testing.s3mock.store.BucketMetadata.objects()" because "bucketMetadata" is null
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:659) ~[spring-beans-6.0.9.jar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:647) ~[spring-beans-6.0.9.jar!/:6.0.
9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1332)
~[spring-beans-6.0.9.jar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1162) ~[spring-be
ans-6.0.9.jar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:560) ~[spring-beans-6.0
.9.jar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:520) ~[spring-beans-6.0.9
.jar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:326) ~[spring-beans-6.0.9.jar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234) ~[spring-beans-6.0.9.jar!/:6.0
.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:324) ~[spring-beans-6.0.9.jar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:200) ~[spring-beans-6.0.9.jar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:254) ~[spring-beans-6.0.9.jar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1417) ~[spring-beans-6.0.9.jar!/
:6.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1337) ~[spring-beans-6.0.9.jar!/:6
.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:888) ~[spring-beans-6.0.9.jar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:791) ~[spring-beans-6.0.9.jar!/:6.0.9]
s3mock-s3mock-1  |      ... 42 common frames omitted
s3mock-s3mock-1  | Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.adobe.testing.s3mock.store.BucketStore]: Factory method 'bucketStore'
 threw exception with message: Cannot invoke "com.adobe.testing.s3mock.store.BucketMetadata.objects()" because "bucketMetadata" is null
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:171) ~[spring-beans-6.0.9.jar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:655) ~[spring-beans-6.0.9.jar!/:6.0.9]
s3mock-s3mock-1  |      ... 56 common frames omitted
s3mock-s3mock-1  | Caused by: java.lang.NullPointerException: Cannot invoke "com.adobe.testing.s3mock.store.BucketMetadata.objects()" because "bucketMetadata" is null
s3mock-s3mock-1  |      at com.adobe.testing.s3mock.store.BucketStore.loadBuckets(BucketStore.java:285) ~[classes!/:na]
s3mock-s3mock-1  |      at com.adobe.testing.s3mock.store.StoreConfiguration.bucketStore(StoreConfiguration.java:72) ~[classes!/:na]
s3mock-s3mock-1  |      at com.adobe.testing.s3mock.store.StoreConfiguration$$SpringCGLIB$$0.CGLIB$bucketStore$1(<generated>) ~[classes!/:na]
s3mock-s3mock-1  |      at com.adobe.testing.s3mock.store.StoreConfiguration$$SpringCGLIB$$2.invoke(<generated>) ~[classes!/:na]
s3mock-s3mock-1  |      at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:258) ~[spring-core-6.0.9.jar!/:6.0.9]
s3mock-s3mock-1  |      at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:331) ~[spring-context-6
.0.9.jar!/:6.0.9]
s3mock-s3mock-1  |      at com.adobe.testing.s3mock.store.StoreConfiguration$$SpringCGLIB$$0.bucketStore(<generated>) ~[classes!/:na]
s3mock-s3mock-1  |      at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
s3mock-s3mock-1  |      at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source) ~[na:na]
s3mock-s3mock-1  |      at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[na:na]
s3mock-s3mock-1  |      at java.base/java.lang.reflect.Method.invoke(Unknown Source) ~[na:na]
s3mock-s3mock-1  |      at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:139) ~[spring-beans-6.0.9.jar!/:6.0.9]
s3mock-s3mock-1  |      ... 57 common frames omitted
s3mock-s3mock-1  |
s3mock-s3mock-1  |
s3mock-s3mock-1 exited with code 1
```
</details>

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -13,6 +13,7 @@
 * [CURRENT - 3.x - THIS VERSION IS UNDER ACTIVE DEVELOPMENT](#current---3x---this-version-is-under-active-development)
   * [3.2.0 - PLANNED](#320---planned)
   * [3.1.0 - PLANNED](#310---planned)
+  * [3.0.1](#301)
   * [3.0.0](#300)
 * [DEPRECATED - 2.x](#deprecated---2x)
   * [2.17.0](#2170)
@@ -129,6 +130,13 @@ Version 3.x is JDK17 LTS bytecode compatible, with Docker and JUnit / direct Jav
   * TBD
 * Version updates
   * TBD
+
+## 3.0.1
+3.x is JDK17 LTS bytecode compatible, with Docker and JUnit / direct Java integration.
+
+* Features and fixes
+  * Fix startup on existing data folder issues (fixes #1245)
+  * Return checksumAlgorithm in ListObjects / ListObjectsV2 (fixes #1220)
 
 ## 3.0.0
 3.x is JDK17 LTS bytecode compatible, with Docker and JUnit / direct Java integration.

--- a/server/src/main/java/com/adobe/testing/s3mock/store/BucketStore.java
+++ b/server/src/main/java/com/adobe/testing/s3mock/store/BucketStore.java
@@ -122,12 +122,14 @@ public class BucketStore {
   public List<UUID> lookupKeysInBucket(String prefix, String bucketName) {
     var bucketMetadata = getBucketMetadata(bucketName);
     var normalizedPrefix = prefix == null ? "" : prefix;
-    return bucketMetadata.objects()
-        .entrySet()
-        .stream()
-        .filter(entry -> entry.getKey().startsWith(normalizedPrefix))
-        .map(Map.Entry::getValue)
-        .toList();
+    synchronized (lockStore.get(bucketName)) {
+      return bucketMetadata.objects()
+          .entrySet()
+          .stream()
+          .filter(entry -> entry.getKey().startsWith(normalizedPrefix))
+          .map(Map.Entry::getValue)
+          .toList();
+    }
   }
 
   /**
@@ -281,11 +283,13 @@ public class BucketStore {
     for (String bucketName : bucketNames) {
       LOG.info("Loading existing bucket {}.", bucketName);
       lockStore.putIfAbsent(bucketName, new Object());
-      var bucketMetadata = getBucketMetadata(bucketName);
-      var objects = bucketMetadata.objects();
-      for (Map.Entry<String, UUID> objectEntry : objects.entrySet()) {
-        objectIds.add(objectEntry.getValue());
-        LOG.info("Loading existing bucket {} key {}", bucketName, objectEntry.getKey());
+      synchronized (lockStore.get(bucketName)) {
+        var bucketMetadata = getBucketMetadata(bucketName);
+        var objects = bucketMetadata.objects();
+        for (Map.Entry<String, UUID> objectEntry : objects.entrySet()) {
+          objectIds.add(objectEntry.getValue());
+          LOG.info("Loading existing bucket {} key {}", bucketName, objectEntry.getKey());
+        }
       }
     }
     return objectIds;

--- a/server/src/main/java/com/adobe/testing/s3mock/store/BucketStore.java
+++ b/server/src/main/java/com/adobe/testing/s3mock/store/BucketStore.java
@@ -43,7 +43,7 @@ import org.slf4j.LoggerFactory;
 public class BucketStore {
 
   private static final Logger LOG = LoggerFactory.getLogger(BucketStore.class);
-  private static final String BUCKET_META_FILE = "bucketMetadata.json";
+  static final String BUCKET_META_FILE = "bucketMetadata.json";
   /**
    * This map stores one lock object per Bucket name.
    * Any method modifying the underlying file must aquire the lock object before the modification.

--- a/server/src/main/java/com/adobe/testing/s3mock/store/MultipartStore.java
+++ b/server/src/main/java/com/adobe/testing/s3mock/store/MultipartStore.java
@@ -407,10 +407,11 @@ public class MultipartStore {
 
   private boolean createPartsFolder(BucketMetadata bucket, UUID id, String uploadId) {
     var partsFolder = getPartsFolderPath(bucket, id, uploadId).toFile();
-    if (!retainFilesOnExit) {
+    var created = partsFolder.mkdirs();
+    if (created && !retainFilesOnExit) {
       partsFolder.deleteOnExit();
     }
-    return partsFolder.mkdirs();
+    return created;
   }
 
   private Path getPartsFolderPath(BucketMetadata bucket, UUID id, String uploadId) {

--- a/server/src/main/java/com/adobe/testing/s3mock/store/ObjectStore.java
+++ b/server/src/main/java/com/adobe/testing/s3mock/store/ObjectStore.java
@@ -53,11 +53,14 @@ import java.util.UUID;
 import java.util.concurrent.ConcurrentHashMap;
 import javax.xml.stream.XMLStreamException;
 import org.apache.commons.io.FileUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * Stores objects and their metadata created in S3Mock.
  */
 public class ObjectStore {
+  private static final Logger LOG = LoggerFactory.getLogger(ObjectStore.class);
   private static final String META_FILE = "objectMetadata.json";
   private static final String ACL_FILE = "objectAcl.xml";
   private static final String DATA_FILE = "binaryData";
@@ -401,10 +404,15 @@ public class ObjectStore {
   }
 
   void loadObjects(BucketMetadata bucketMetadata, Collection<UUID> ids) {
+    var loaded = 0;
     for (var id : ids) {
       lockStore.putIfAbsent(id, new Object());
-      getS3ObjectMetadata(bucketMetadata, id);
+      var s3ObjectMetadata = getS3ObjectMetadata(bucketMetadata, id);
+      if (s3ObjectMetadata != null) {
+        loaded++;
+      }
     }
+    LOG.info("Loaded {}/{} objects for bucket {}", loaded, ids.size(), bucketMetadata.name());
   }
 
   /**

--- a/server/src/main/java/com/adobe/testing/s3mock/store/ObjectStore.java
+++ b/server/src/main/java/com/adobe/testing/s3mock/store/ObjectStore.java
@@ -460,7 +460,9 @@ public class ObjectStore {
    */
   private void createObjectRootFolder(BucketMetadata bucket, UUID id) {
     var objectRootFolder = getObjectFolderPath(bucket, id).toFile();
-    objectRootFolder.mkdirs();
+    if (objectRootFolder.mkdirs() && !retainFilesOnExit) {
+      objectRootFolder.deleteOnExit();
+    }
   }
 
   private Path getObjectFolderPath(BucketMetadata bucket, UUID id) {

--- a/server/src/main/java/com/adobe/testing/s3mock/store/StoreConfiguration.java
+++ b/server/src/main/java/com/adobe/testing/s3mock/store/StoreConfiguration.java
@@ -25,7 +25,6 @@ import java.time.format.DateTimeFormatter;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
-import java.util.Objects;
 import org.apache.commons.io.FileUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -77,7 +76,7 @@ public class StoreConfiguration {
 
   @Bean
   List<String> bucketNames(File rootFolder) {
-    var buckets = rootFolder.listFiles((File dir, String name) -> !Objects.equals(name, "test"));
+    var buckets = rootFolder.listFiles();
     if (buckets != null) {
       return Arrays.stream(buckets).map(File::getName).toList();
     } else {

--- a/server/src/main/java/com/adobe/testing/s3mock/store/StoreConfiguration.java
+++ b/server/src/main/java/com/adobe/testing/s3mock/store/StoreConfiguration.java
@@ -16,14 +16,15 @@
 
 package com.adobe.testing.s3mock.store;
 
+import static com.adobe.testing.s3mock.store.BucketStore.BUCKET_META_FILE;
+
 import com.fasterxml.jackson.databind.ObjectMapper;
 import java.io.File;
 import java.io.IOException;
 import java.nio.file.Files;
 import java.time.ZoneId;
 import java.time.format.DateTimeFormatter;
-import java.util.Arrays;
-import java.util.Collections;
+import java.util.ArrayList;
 import java.util.List;
 import org.apache.commons.io.FileUtils;
 import org.slf4j.Logger;
@@ -76,12 +77,21 @@ public class StoreConfiguration {
 
   @Bean
   List<String> bucketNames(File rootFolder) {
-    var buckets = rootFolder.listFiles();
-    if (buckets != null) {
-      return Arrays.stream(buckets).map(File::getName).toList();
-    } else {
-      return Collections.emptyList();
+    var bucketNames = new ArrayList<String>();
+    try (var paths = Files.newDirectoryStream(rootFolder.toPath())) {
+      paths.forEach(
+          path ->  {
+            var resolved = path.resolve(BUCKET_META_FILE);
+            if (resolved.toFile().exists()) {
+              bucketNames.add(path.getFileName().toString());
+            }
+          }
+      );
+    } catch (IOException e) {
+      throw new IllegalStateException("Could not load buckets from data directory "
+          + rootFolder, e);
     }
+    return bucketNames;
   }
 
   @Bean

